src/core/Form/SearchInput/SearchInput.tsx

```diff
@@ -1,47 +1,265 @@
-import { TextInputProps, textInputClassNames } from '../TextInput/TextInput';
-import { Icon } from '../../Icon/Icon';
 import classnames from 'classnames';
+import {
+  HtmlInputWithRef,
+  HtmlInputProps,
+  HtmlDiv,
+  HtmlDivProps,
+  HtmlButton,
+  HtmlButtonProps,
+} from '../../../reset';
 import { Omit } from '../../../utils/typescript';
+import { idGenerator } from '../../../utils/uuid';
+import { TokensProp, InternalTokensProp } from '../../theme';
+import { withSuomifiDefaultProps } from '../../theme/utils';
+import { VisuallyHidden } from '../../../components/Visually-hidden/Visually-hidden';
+import { StatusText } from '../StatusText/StatusText';
+import { LabelText, LabelMode } from '../LabelText/LabelText';
+import { Icon } from '../../Icon/Icon';
+import { baseStyles } from './SearchInput.baseStyles';
+
+type StateValue = string | number | string[] | undefined;
+
+type SearchInputStatus = 'default' | 'error';
+
+export interface SearchInputProps
+  extends Omit<
+      HtmlInputProps,
+      'type' | 'disabled' | 'onChange' | 'children' | 'onClick'
+    >,
+    TokensProp {
+  /** SearchInput container div class name for custom styling. */
+  className?: string;
+  /** SearchInput container div props */
+  inputContainerProps?: Omit<HtmlDivProps, 'className'>;
+  /** Label text */
+  labelText: string;
+  /** Hide or show label. Label element is always present, but can be visually hidden.
+   * @default visible
+   */
+  labelMode?: LabelMode;
+  /** Placeholder text for input. Use only as visual aid, not for instructions. */
+  visualPlaceholder?: string;
+  /** Clear button label for screen readers */
+  clearButtonLabel: string;
+  /** Search button label for screen readers */
+  searchButtonLabel: string;
+  /** SearchButtonProps */
+  searchButtonProps?: Omit<HtmlButtonProps, 'onClick' | 'tabIndex'>;
+  /**
+   * 'default' | 'error'
+   * @default default
+   */
+  status?: SearchInputStatus;
+  /** Status text to be shown below the component and hint text. Use e.g. for validation error */
+  statusText?: string;
+  /** Input name */
+  name?: string;
+  /** Set components width to 100% */
+  fullWidth?: boolean;
+  /** Callback for input text change */
+  onChange?: (value: StateValue) => void;
+  /** Callback for onBlur event */
+  onBlur?: (event: FocusEvent<HTMLInputElement>) => void;
+  /** Callback for search button click */
+  onSearch?: (value: StateValue) => void;
+}
 
 const baseClassName = 'fi-search-input';
-const iconBaseClassName = `${baseClassName}_icon`;
+const searchInputClassNames = {
+  fullWidth: `${baseClassName}--full-width`,
+  error: `${baseClassName}--error`,
+  notEmpty: `${baseClassName}--not-empty`,
+  inputElement: `${baseClassName}_input`,
+  inputElementContainer: `${baseClassName}_input-element-container`,
+  functionalityContainer: `${baseClassName}_functionality-container`,
+  button: `${baseClassName}_button`,
```

Comment:
Do buttons need this common classname? 
Could it then also be nice to have one common for icons, like `button-icon`?

Reply:
Shared button styles are applied this way, I guess it would be possible to arrange the styles in a different way as well by extending. Icons do not share styles so common classname would only serve the purpose of custom styling which in my opinion should not be allowed in this case.

Reply:
Okay, it sounds good to leave as it is now.